src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java

```diff
@@ -331,27 +340,67 @@ public InputStream executeCommand(String cmd, Object ... args) throws IOExceptio
     protected static final int VERSION_1 = 1;
     protected static final int VERSION_2 = 2;
 
+    // Attach operation properties.
+    protected static class OperationProperties {
+        public final static String STREAMING = "streaming";
+
+        private int ver;
+        private Map<String, String> options = new HashMap<>();
+
+        OperationProperties(int ver) {
+            this.ver = ver;
+        }
+
+        int version() {
+            return ver;
+        }
+
+        void setOption(String name, String value) {
+            options.put(name, value);
+        }
+
+        String options() {
+            return options.entrySet().stream()
+                          .map(e -> e.getKey() + "=" + e.getValue())
+                          .collect(Collectors.joining(","));
+        }
+    }
+
     /*
-     * Detects Attach API version supported by target VM.
+     * Detects Attach API properties supported by target VM.
      */
-    protected int detectVersion() throws IOException {
+    protected OperationProperties getDefaultProps() throws IOException {
         try {
-            InputStream reply = execute("getversion");
+            InputStream reply = execute("getversion", "options");
             String message = readMessage(reply);
             reply.close();
-            try {
-                int supportedVersion = Integer.parseUnsignedInt(message);
-                // we expect only VERSION_2
-                if (supportedVersion == VERSION_2) {
-                    return VERSION_2;
+
+            // Reply is "<ver> option1,option2...".
+            int delimPos = message.indexOf(' ');
```

Comment:
We can do this in an early-break sort of way. That'd reduce the indentation and let's us not to have to check `delimPos` again.

```java
int delimPos = message.indexOf(' ');
int supportedVersion = Integer.parseUnsignedInt(delimPos < 0 ? message : message.substring(0, delimPos));
if (delimPos < 0 || supportedVersion != VERSION_2) {
    return; // No options to parse, or doesn't support options anyway
}
/* Insert the parsing stuff here */
```

Reply:
To me current implementation looks clearer. And I think it will be simpler to extend the logic (if we ever need to extend v2 or introduce v3)